### PR TITLE
Add Qos case

### PR DESCRIPTION
**Pull request by chloerh**
- Update the QoS setting in interface by update-device

Signed-off-by: haizhao <haizhao@redhat.com>

### CODE REVIEW

**Comment by chloerh** [comment]
(1/1) type_specific.local.virtual_network.iface_network.qos_test.qos_option_bandwidth.by_update_device: PASS (29.50 s)

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 1]
```diff
@@ -933,6 +955,24 @@ def run_guest_libvirt(session):
...
                 if re.search(r"%s" % ipt, output_des, re.M):
                     test.fail("Find ip6table rule %s after net destroyed" % ipt)
+        if remove_bandwidth:
+            iface_name = libvirt.get_ifname_host(vm_name, iface_mac)
+            cur_xml = virsh.dumpxml(vm_name).stdout_text
+            logging.debug(cur_xml)
+            if 'bandwidth' in cur_xml:
+                test.fail('bandwidth still in xml')
+            if not check_filter_rules(iface_name, 0, expect_none=True):
+                test.fail('Expect nothing in output')
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_network/iface_network.py]: I can not get the failure reason with the msg, can to explain more?

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 2]
```diff
@@ -67,7 +67,7 @@ def prepare_pxe_boot():
         with open(pxe_file, 'w') as p_file:
             p_file.write(boot_txt)
 
-    def modify_iface_xml():
+    def modify_iface_xml(sync=True):
         """
         Modify interface xml options
         """
```
Comment (kylazhang) [inline libvirt/tests/src/virtual_network/iface_network.py]: docstring needed here for parameters

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 3]
```diff
@@ -663,7 +673,11 @@ def run_guest_libvirt(session):
         # Edit the interface xml.
         if change_iface_option:
             try:
-                modify_iface_xml()
+                if update_device:
+                    updated_iface = modify_iface_xml(sync=False)
+                    virsh.update_device(vm_name, updated_iface.xml, debug=True)
```
[the remark this answers is not in the document]
Comment (chloerh) [reply libvirt/tests/src/virtual_network/iface_network.py]: Updated

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 4]
```diff
@@ -787,15 +801,23 @@ def run_guest_libvirt(session):
                 run_dnsmasq_addnhosts_test(net_dns_hostip, net_dns_hostnames)
 
         # Run bandwidth test for network
-        if test_qos_bandwidth:
+        if test_qos_bandwidth and not update_device:
             run_bandwidth_test(check_net=True)
+
+        #
```
[the remark this answers is not in the document]
Comment (chloerh) [reply libvirt/tests/src/virtual_network/iface_network.py]: Updated

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 5]
```diff
@@ -933,6 +955,24 @@ def run_guest_libvirt(session):
...
+            if 'bandwidth' in cur_xml:
+                test.fail('bandwidth still in xml')
+            if not check_filter_rules(iface_name, 0, expect_none=True):
+                test.fail('Expect nothing in output')
+        if update_device and loop:
+            loop -= 1
+            if loop:
+                # Rerun this procedure again with updated params
+                # Reset params of the corresponding loop
+                loop_sufix = 'loop' + str(loop) + '_'
```
[the remark this answers is not in the document]
Comment (chloerh) [reply libvirt/tests/src/virtual_network/iface_network.py]: Updated

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 6]
```diff
@@ -787,15 +801,23 @@ def run_guest_libvirt(session):
...
         # Run bandwidth test for network
-        if test_qos_bandwidth:
+        if test_qos_bandwidth and not update_device:
             run_bandwidth_test(check_net=True)
+
+        #
+        if remove_bandwidth:
+            ori_iface = params['original_iface']
+            logging.debug(ori_iface)
+            virsh.update_device(vm_name, ori_iface.xml, debug=True)
```
[the remark this answers is not in the document]
Comment (chloerh) [reply libvirt/tests/src/virtual_network/iface_network.py]: Updated

File: libvirt/tests/src/virtual_network/iface_network.py [hunk 7]
```diff
@@ -222,13 +227,15 @@ def check_class_rules(ifname, rule_id, bandwidth):
                 tc_burst = int(se.group(5))
             assert tc_burst == int(bandwidth["burst"])
 
-    def check_filter_rules(ifname, bandwidth):
+    def check_filter_rules(ifname, bandwidth, expect_none=False):
```
[the remark this answers is not in the document]
Comment (chloerh) [reply libvirt/tests/src/virtual_network/iface_network.py]: Updated, but there's no return in this function.

File: libvirt/tests/src/virtual_network/iface_network.py (hunk 1 above)
Comment (chloerh) [reply libvirt/tests/src/virtual_network/iface_network.py]: Updated

File: libvirt/tests/src/virtual_network/iface_network.py (hunk 2 above)
Comment (chloerh) [reply libvirt/tests/src/virtual_network/iface_network.py]: Updated

File: libvirt/tests/src/virtual_network/iface_network.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (kylazhang) [reply libvirt/tests/src/virtual_network/iface_network.py]: It have, on line 246

**Comment by chloerh** [comment]
Test result before this commit:
---
 (01/35) type_specific.local.virtual_network.iface_network.rom_test.negative.enable_no_bar_off: PASS (6.25 s)
 (02/35) type_specific.local.virtual_network.iface_network.rom_test.negative.enable_no_with_file: PASS (7.35 s)
 (03/35) type_specific.local.virtual_network.iface_network.bridge_test: PASS (21.26 s)
 (04/35) type_specific.local.virtual_network.iface_network.dnsmasq_test.net_dns_forward: PASS (18.86 s)
 (05/35) type_specific.local.virtual_network.iface_network.dnsmasq_test.net_domain: PASS (21.66 s)
 (06/35) type_specific.local.virtual_network.iface_network.dnsmasq_test.net_host_ip: ERROR: Timeout expired while waiting for shell command to complete: 'yum install -y bind-utils'    (output: "Loaded plugins: product-id, search-disabled-repos, subscription-manager\nThis system is not registered with an entitlement server. You can use subscripti... (553.54 s)
 (07/35) type_specific.local.virtual_network.iface_network.dnsmasq_test.net_dhcp_range: FAIL: VM failed to start:\nVM 'avocado-vt-vm1_0' failed to start: error: failed to get domain 'avocado-vt-vm1_0'\nerror: Domain not found: no domain with matching name 'avocado-vt-vm1_0' (25.78 s)
 (08/35) type_specific.local.virtual_network.iface_network.dnsmasq_test.net_dns_host: PASS (22.76 s)
 (09/35) type_specific.local.virtual_network.iface_network.qos_test.qos_option_bandwidth.positive_test: PASS (22.88 s)
 (10/35) type_specific.local.virtual_network.iface_network.qos_test.qos_option_bandwidth.negtive_start_test: PASS (5.54 s)
 (11/35) type_specific.local.virtual_network.iface_network.qos_test.qos_option_bandwidth.negtive_remove_test: PASS (20.40 s)
 (12/35) type_specific.local.virtual_network.iface_network.net_forward.net_isolated: PASS (40.91 s)
 (13/35) type_specific.local.virtual_network.iface_network.net_forward.net_nat: PASS (38.39 s)
 (14/35) type_specific.local.virtual_network.iface_network.net_forward.net_route: PASS (37.95 s)
 (15/35) type_specific.local.virtual_network.iface_network.net_forward.net_open.default: PASS (30.34 s)
 (16/35) type_specific.local.virtual_network.iface_network.net_forward.net_open.without_bridge: PASS (28.52 s)
 (17/35) type_specific.local.virtual_network.iface_network.net_forward.net_open.without_mac: PASS (29.33 s)
 (18/35) type_specific.local.virtual_network.iface_network.net_forward.net_open.negative_without_ip: PASS (4.50 s)
 (19/35) type_specific.local.virtual_network.iface_network.net_forward.net_open.negative_with_dev: PASS (4.88 s)
 (20/35) type_specific.local.virtual_network.iface_network.net_forward.route_test: PASS (39.88 s)
 (21/35) type_specific.local.virtual_network.iface_network.net_pxe.default: CANCEL: Please provide initrd/vmlinuz URL (3.86 s)
 (22/35) type_specific.local.virtual_network.iface_network.net_pxe.boot_failure.virtio: CANCEL: Please provide initrd/vmlinuz URL (3.78 s)
 (23/35) type_specific.local.virtual_network.iface_network.net_pxe.boot_failure.e1000: CANCEL: Please provide initrd/vmlinuz URL (3.75 s)
 (24/35) type_specific.local.virtual_network.iface_network.net_pxe.boot_failure.rtl8139: CANCEL: Please provide initrd/vmlinuz URL (4.56 s)
 (25/35) type_specific.local.virtual_network.iface_network.net_pxe.boot_failure.e1000e: CANCEL: Please provide initrd/vmlinuz URL (3.85 s)
 (26/35) type_specific.local.virtual_network.iface_network.net_guest: ERROR: Timeout expired while waiting for shell command to complete: 'yum install -y libvirt'    (output: "Loaded plugins: product-id, search-disabled-repos, subscription-manager\nThis system is not registered with an entitlement server. You can use subscription-... (551.72 s)
 (27/35) type_specific.local.virtual_network.iface_network.net_macvtap.error_test.start_error: PASS (5.20 s)
 (28/35) type_specific.local.virtual_network.iface_network.net_macvtap.error_test.define_error: PASS (4.63 s)
 (29/35) type_specific.local.virtual_network.iface_network.net_macvtap.net_bridge.default_test: PASS (25.40 s)
 (30/35) type_specific.local.virtual_network.iface_network.net_macvtap.net_bridge.multi_guests: FAIL: VM failed to start:\nVM 'avocado-vt-vm1_0' failed to start: error: failed to get domain 'avocado-vt-vm1_0'\nerror: Domain not found: no domain with matching name 'avocado-vt-vm1_0' (24.45 s)
 (31/35) type_specific.local.virtual_network.iface_network.net_macvtap.net_passthrough: PASS (63.11 s)
 (32/35) type_specific.local.virtual_network.iface_network.net_macvtap.net_private: PASS (23.07 s)
 (33/35) type_specific.local.virtual_network.iface_network.net_portgroup.default_portgroup: PASS (27.10 s)
 (34/35) type_specific.local.virtual_network.iface_network.net_portgroup.specific_portgroup: PASS (26.77 s)
 (35/35) type_specific.local.virtual_network.iface_network.net_portgroup.overwriting_bandwidth: PASS (26.78 s)
RESULTS    : PASS 26 | ERROR 2 | FAIL 2 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 5
JOB TIME   : 1788.74 s


Test result after this commit
---
 (01/36) type_specific.local.virtual_network.iface_network.rom_test.negative.enable_no_bar_off: PASS (6.21 s)
 (02/36) type_specific.local.virtual_network.iface_network.rom_test.negative.enable_no_with_file: PASS (5.61 s)
 (03/36) type_specific.local.virtual_network.iface_network.bridge_test: PASS (18.71 s)
 (04/36) type_specific.local.virtual_network.iface_network.dnsmasq_test.net_dns_forward: PASS (21.37 s)
 (05/36) type_specific.local.virtual_network.iface_network.dnsmasq_test.net_domain: PASS (20.05 s)
 (06/36) type_specific.local.virtual_network.iface_network.dnsmasq_test.net_host_ip: ERROR: Timeout expired while waiting for shell command to complete: 'yum install -y bind-utils'    (output: "Loaded plugins: product-id, search-disabled-repos, subscription-manager\nThis system is not registered with an entitlement server. You can use subscripti... (553.78 s)
 (07/36) type_specific.local.virtual_network.iface_network.dnsmasq_test.net_dhcp_range: FAIL: VM failed to start:\nVM 'avocado-vt-vm1_0' failed to start: error: failed to get domain 'avocado-vt-vm1_0'\nerror: Domain not found: no domain with matching name 'avocado-vt-vm1_0' (26.02 s)
 (08/36) type_specific.local.virtual_network.iface_network.dnsmasq_test.net_dns_host: PASS (23.85 s)
 (09/36) type_specific.local.virtual_network.iface_network.qos_test.qos_option_bandwidth.positive_test: PASS (23.64 s)
 (10/36) type_specific.local.virtual_network.iface_network.qos_test.qos_option_bandwidth.by_update_device: PASS (31.17 s)
 (11/36) type_specific.local.virtual_network.iface_network.qos_test.qos_option_bandwidth.negtive_start_test: PASS (5.44 s)
 (12/36) type_specific.local.virtual_network.iface_network.qos_test.qos_option_bandwidth.negtive_remove_test: PASS (20.44 s)
 (13/36) type_specific.local.virtual_network.iface_network.net_forward.net_isolated: PASS (40.14 s)
 (14/36) type_specific.local.virtual_network.iface_network.net_forward.net_nat: PASS (37.49 s)
 (15/36) type_specific.local.virtual_network.iface_network.net_forward.net_route: PASS (37.35 s)
 (16/36) type_specific.local.virtual_network.iface_network.net_forward.net_open.default: PASS (31.19 s)
 (17/36) type_specific.local.virtual_network.iface_network.net_forward.net_open.without_bridge: PASS (28.73 s)
 (18/36) type_specific.local.virtual_network.iface_network.net_forward.net_open.without_mac: PASS (28.82 s)
 (19/36) type_specific.local.virtual_network.iface_network.net_forward.net_open.negative_without_ip: PASS (4.51 s)
 (20/36) type_specific.local.virtual_network.iface_network.net_forward.net_open.negative_with_dev: PASS (4.62 s)
 (21/36) type_specific.local.virtual_network.iface_network.net_forward.route_test: PASS (32.87 s)
 (22/36) type_specific.local.virtual_network.iface_network.net_pxe.default: CANCEL: Please provide initrd/vmlinuz URL (14.37 s)
 (23/36) type_specific.local.virtual_network.iface_network.net_pxe.boot_failure.virtio: CANCEL: Please provide initrd/vmlinuz URL (3.66 s)
 (24/36) type_specific.local.virtual_network.iface_network.net_pxe.boot_failure.e1000: CANCEL: Please provide initrd/vmlinuz URL (3.66 s)
 (25/36) type_specific.local.virtual_network.iface_network.net_pxe.boot_failure.rtl8139: CANCEL: Please provide initrd/vmlinuz URL (3.64 s)
 (26/36) type_specific.local.virtual_network.iface_network.net_pxe.boot_failure.e1000e: CANCEL: Please provide initrd/vmlinuz URL (3.65 s)
 (27/36) type_specific.local.virtual_network.iface_network.net_guest: ERROR: Timeout expired while waiting for shell command to complete: 'yum install -y libvirt'    (output: "Loaded plugins: product-id, search-disabled-repos, subscription-manager\nThis system is not registered with an entitlement server. You can use subscription-... (552.67 s)
 (28/36) type_specific.local.virtual_network.iface_network.net_macvtap.error_test.start_error: PASS (4.89 s)
 (29/36) type_specific.local.virtual_network.iface_network.net_macvtap.error_test.define_error: PASS (4.60 s)
 (30/36) type_specific.local.virtual_network.iface_network.net_macvtap.net_bridge.default_test: PASS (24.34 s)
 (31/36) type_specific.local.virtual_network.iface_network.net_macvtap.net_bridge.multi_guests: FAIL: VM failed to start:\nVM 'avocado-vt-vm1_0' failed to start: error: failed to get domain 'avocado-vt-vm1_0'\nerror: Domain not found: no domain with matching name 'avocado-vt-vm1_0' (24.02 s)
 (32/36) type_specific.local.virtual_network.iface_network.net_macvtap.net_passthrough: PASS (63.21 s)
 (33/36) type_specific.local.virtual_network.iface_network.net_macvtap.net_private: PASS (23.68 s)
 (34/36) type_specific.local.virtual_network.iface_network.net_portgroup.default_portgroup: PASS (27.79 s)
 (35/36) type_specific.local.virtual_network.iface_network.net_portgroup.specific_portgroup: PASS (26.90 s)
 (36/36) type_specific.local.virtual_network.iface_network.net_portgroup.overwriting_bandwidth: PASS (27.05 s)
RESULTS    : PASS 27 | ERROR 2 | FAIL 2 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 5
JOB TIME   : 1820.11 s


As it shows, 1 case added, and nothing else is changed.

**Comment by kylazhang** [comment]
Thanks for the result

**Pull request merged**